src/beatree/allocator/mod.rs

```diff
@@ -58,56 +58,26 @@ pub struct AllocatorWriter {
     released: Vec<PageNumber>,
 }
 
-/// creates a pair of AllocatorReader and AllocatorWriter over a possibly already existing File.
-pub fn create(
-    fd: File,
-    free_list_head: Option<PageNumber>,
-    bump: PageNumber,
-    wr_io_handle_index: usize,
-    wr_io_sender: Sender<IoCommand>,
-    wr_io_receiver: Receiver<CompleteIo>,
-    rd_io_handle_index: usize,
-    rd_io_sender: Sender<IoCommand>,
-    rd_io_receiver: Receiver<CompleteIo>,
-) -> (AllocatorReader, AllocatorWriter) {
-    let file_size = fd
-        .metadata()
-        .expect("Error extracting metadata from file")
-        .size() as usize;
-
-    let reader_fd = fd.try_clone().expect("failed to clone file");
-    let writer_fd = fd;
-
-    let writer = AllocatorWriter {
-        free_list: FreeList::read(
-            &writer_fd,
-            &wr_io_sender,
-            wr_io_handle_index,
-            &wr_io_receiver,
+impl AllocatorReader {
+    /// creates an AllocatorReader over a possibly already existing File.
+    pub fn new(
+        fd: File,
+        free_list_head: Option<PageNumber>,
+        bump: PageNumber,
+        io_handle_index: usize,
+        io_sender: Sender<IoCommand>,
+        io_receiver: Receiver<CompleteIo>,
+    ) -> Self {
+        AllocatorReader {
+            store_file: fd,
             free_list_head,
```

Comment:
Just a fly by comment: there is no reason for free list head to be in the allocator reader.

Reply:
The `AllocatorReader` is able to parse the entire free list from the file, so it needs to know where the head is. I agree that a refactor of the code could place the free list parsing somewhere else to avoid keeping the head in the `AllocatorReader`

Reply:
exactly